decoder/stats.go

```diff
@@ -375,6 +422,154 @@ func updatePokemonStats(old *Pokemon, new *Pokemon, areas []geo.AreaName) {
 	}
 }
 
+func updateRaidStats(old *Gym, new *Gym, areas []geo.AreaName) {
+	if len(areas) == 0 {
+		areas = []geo.AreaName{
+			{
+				Parent: "world",
+				Name:   "world",
+			},
+		}
+	}
+
+	locked := false
+
+	// Loop though all areas
+	for i := 0; i < len(areas); i++ {
+		area := areas[i]
+
+		// Check if Raid has started/is active or RaidEndTimestamp has changed (Back-to-back raids)
+		if new.RaidPokemonId.ValueOrZero() > 0 &&
+		(old == nil || old.RaidPokemonId != new.RaidPokemonId || old.RaidEndTimestamp != new.RaidEndTimestamp) {
+
+			if !locked {
+				raidStatsLock.Lock()
+				locked = true
+			}
+
+			countStats := raidCount[area]
+			if countStats == nil {
+				countStats = make(map[int64]areaRaidCountDetail)
+				raidCount[area] = countStats
+			}
+
+			var countRaids = raidCount[area][new.RaidLevel.ValueOrZero()]
+			countRaids.count[new.RaidPokemonId.ValueOrZero()]++
+			raidCount[area][new.RaidLevel.ValueOrZero()] = countRaids
+		}
+	}
+
+	if locked {
+		raidStatsLock.Unlock()
+	}
+}
+
+func updateIncidentStats(old *Incident, new *Incident, areas []geo.AreaName) {
+	if len(areas) == 0 {
+		areas = []geo.AreaName{
+			{
+				Parent: "world",
+				Name:   "world",
+			},
+		}
+	}
+
+	locked := false
+
+	// Loop though all areas
+	for i := 0; i < len(areas); i++ {
+		area := areas[i]
+
+		// Check if StartTime has changed, then we can assume a new Incident has appeared.
+		if old == nil || old.StartTime != new.StartTime {
+
+			if !locked {
+				incidentStatsLock.Lock()
+				locked = true
+			}
+
+			invasionStats := invasionCount[area]
+			if invasionStats == nil {
+				invasionStats = &areaInvasionCountDetail{}
+				invasionCount[area] = invasionStats
+			}
+
+			// Exclude Kecleon, Showcases and other UNSET characters for invasionStats.
+			if new.Character != 0 {
+				invasionStats.count[new.Character]++
+			}
+		}
+	}
+
+	if locked {
+		incidentStatsLock.Unlock()
+	}
+}
+
+func updateQuestStats(pokestop *Pokestop, haveAr bool, areas []geo.AreaName) {
```

Comment:
I think in this is still a logical mistake. a pokestop is updated during day on different events like lure, invasion, powerup_points etc.

In this method you aren't checking for any differences on the quests. so you could count a quest several times during the whole day when events on the stop are happening.

Reply:
hmm maybe my bad.. followed the code where you call it.. you call it in quest update method...... I wonder why I have quest data for today then, as I tested this branch around 9:45